go/oasis-test-runner/scenario/e2e/runtime/runtime_upgrade.go

```diff
@@ -106,7 +109,72 @@ func (sc *runtimeUpgradeImpl) Run(ctx context.Context, childEnv *env.Env) error
 	// Run client again.
 	sc.Logger.Info("starting a second client to check if runtime works")
 	sc.Scenario.TestClient = NewTestClient().WithSeed("seed2").WithScenario(InsertRemoveEncWithSecretsScenarioV2)
-	return sc.RunTestClientAndCheckLogs(ctx, childEnv)
+	if err := sc.RunTestClientAndCheckLogs(ctx, childEnv); err != nil {
+		return err
+	}
+
+	// Ensure that after upgrade, every compute worker had its old exploded
+	// bundle removed from its bundles dir.
+	// We do so by getting the manifest hash of the active deployment,
+	// and ensuring there is a dir with corresponding hash.
+	manifestHash, err := sc.activeDeploymentManifestHash(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch active deployment manifest hash: %w", err)
+	}
+	for _, worker := range sc.Net.ComputeWorkers() {
+		if err := sc.verifyBundleDir(worker, manifestHash); err != nil {
+			return fmt.Errorf("failed to verify clean-up of %s's bundle dir after upgrade: %w", worker.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (sc *runtimeUpgradeImpl) verifyBundleDir(worker *oasis.Compute, manifestHash hash.Hash) error {
+	sc.Logger.Info("ensuring cached exploded bundle for old version was removed",
+		"worker", worker.Name,
+	)
+	// There should be only one exploded bundle dir.
+	bundlesDir := bundle.ExplodedPath(worker.DataDir())
+	entries, err := os.ReadDir(bundlesDir)
+	if err != nil {
+		return err
+	}
+	if n := len(entries); n != 1 {
+		return fmt.Errorf("unexpected number of dir entries: expected 1, got %d", n)
+	}
+	entry := entries[0]
+	if !entry.IsDir() {
+		return fmt.Errorf("%s is not a dir", entry.Name())
+	}
+	// Ensure single entry name (exploded bundle dir) equals to manifest hash.
+	if entry.Name() != manifestHash.String() {
+		return fmt.Errorf("unexpected folder name: want %v, got %v", entry.Name(), manifestHash.String())
+	}
+	return nil
+}
+
+func (sc *runtimeUpgradeImpl) activeDeploymentManifestHash(ctx context.Context) (hash.Hash, error) {
+	var manifestHash hash.Hash
+	rt, err := sc.Net.Controller().Registry.GetRuntime(ctx, &registry.GetRuntimeQuery{
+		Height: consensus.HeightLatest,
+		ID:     KeyValueRuntimeID,
+	})
+	if err != nil {
+		return manifestHash, fmt.Errorf("failed to get runtime descriptor: %w", err)
```

Comment:
```suggestion
		return hash.Hash{}, fmt.Errorf("failed to get runtime descriptor: %w", err)
```
This way it is immediately obvious that you are returning a zero value on error.